retry/api.py

```diff
@@ -1,84 +1,215 @@
+import asyncio
 import logging
 import random
-import time
-
+import traceback
+import threading
 from functools import partial
 
-from .compat import decorator
+import inspect
 
+from .compat import decorator
 
 logging_logger = logging.getLogger(__name__)
 
+class ConditionRelease(object):
+    def __init__(self, c):
+        self.c = c
+    def __enter__(self):
+        self.c.acquire()
+        return self
+    def wait(self, n):
+        self.c.wait(n)
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        self.c.release()
```

Comment:
Can a `contextlib.contextmanager` be used instead? Would it make it simpler?

Reply:
Will change the context manager class, but contextmanager decorator alone is not something that seems fullproof enough.  This context manager class will make code easier to read.